File: 1570-FinalPricesWithASpecialDiscountInAShop/1570-FinalPricesWithASpecialDiscountInAShop.py
# Last updated: 6/17/2025, 10:25:42 PM
class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        result = prices[:]  
        stack = []  

        for i in range(len(prices)):
            while stack and prices[stack[-1]] >= prices[i]:
                index = stack.pop()  
                result[index] -= prices[i]  
            
            stack.append(i) 
        
        return result 

File: 1570-FinalPricesWithASpecialDiscountInAShop/1570-FinalPricesWithASpecialDiscountInAShop.java
// Last updated: 8/23/2025, 6:49:47 PM
class Solution {
    public int[] finalPrices(int[] prices) {
        // Optimal Solution - Monotonic Increasing Stack
       int n = prices.length;
       int[] result = prices.clone();
       Stack<Integer> stack = new Stack<>();

       // Iterating through all elements 
       for(int i = 0; i < n; i++){
        while(!stack.isEmpty() && prices[stack.peek()] >= prices[i]){
            int index = stack.pop();
            result[index] = result[index] - prices[i];

        }
        stack.push(i);
       }
       return result;
        
    }
}